### ISSUE TITLE
DynamicDataSource caching causes problems when running tests in parallel.

### ISSUE THREAD

**Issue by Haplois** [microsoft/testfx#906]
> After 2.2.4, _as long as the values in data source are serializable_, every row for the specific test case is considered a separate test case. If our assumptions break parallel test runs that's a regression. Is it possible for you to share a quick repro with us?

@Haplois I will share a complete project when I have more time, but in general, consider a test case like this:
```cs
[
  DataTestMethod,
  DynamicData(dynamicDataSourceName: nameof(GetData), DynamicDataSourceType.Method)
]
public async Task Create_Tasks_Requests_From_Invalid_Tasks_Definitions_Should_Return_BadRequest_Errors(Events.APIGatewayProxyRequest request, Requests.CreateTaskCommand.Arguments arguments)
{
    // test code
}
```
with `GetData` similar to this:
```cs
public static IEnumerable<object[]> GetData() =>
 new [] {
    new [] object {
         // initialize pair of objects
    },
    new [] object {
         // initialize pair of objects
    },
 }
```
If I run such test cases in parallel (method level parallelism)  then I see all kinds of exception suggesting that the the test framework attempts to pull the enumerable on different threads. I have plenty of data sources structured like this:
```cs
public static IEnumerable<object[]> GetData()
{
    yield new InputData {
       // ...
    };

    yield new InputData {
       // ...
    };

    // etc.
}
```
These obviously can't be pulled in parallel either.

_Originally posted by @paul-michalik in https://github.com/microsoft/testfx/issues/871#issuecomment-874630277_

[AB#1408168](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1408168)

**Comment by Haplois** [comment]
This class needs to be able work concurrently.

https://github.com/microsoft/testfx/blob/main/src/Adapter/MSTest.CoreAdapter/Helpers/DataSerializationHelper.cs#L16

**Comment by jr01** [comment]
@Haplois - just to add to this. I intermittently get this exception with 2.2.5:
```
  Message: 
    Object reference not set to an instance of an object.

  Stack Trace: 
    <>c__DisplayClass4_0.<GetSerializer>b__0(KeyValuePair`2 i)
    Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
    DataSerializationHelper.GetSerializer(String typeName)
    DataSerializationHelper.Deserialize(String[] serializedData)
    TestMethodRunner.RunTestMethod()
    TestMethodRunner.Execute()
```

Changing the `SerializerCache` to a `ConcurrentDictionary` should fix this.

**Comment by Haplois** [comment]
Fixed by #998
